### PR TITLE
kernel: enable config CONFIG_MEMORY_HOTPLUG_DEFAULT_ONLINE by default

### PR DESCRIPTION
**Pull request by wenlxie**
In my test env, I assigned 120Gi memory to the container, but found that Sandbox can't be launched. For details: https://github.com/kata-containers/runtime/issues/1244
kata-agent and udev process was killed because of OOM triggerd.
https://lwn.net/Articles/668944/

Signed-off-by:  Wenli Xie  wenlxie@ebay.com

### CODE REVIEW

**Comment by grahamwhaley** [comment]
:-) I think more importantly is how this does or does not fit in with the memory hotplug flow of kata - @jcvenegas @devimc for input from the 'how is kata hotplug meant to work' side.
But, this is also very good evidence for the 120Gb hotplug failure case as well - nice find @wenlxie

**Comment by devimc** [comment]
@wenlxie great job! do you know why kata-agent receives OOM?
Currently we connect the memory (and CPUs) manually through gRPC/kata-agent, with this patch this implementation should change. I'm not sure but I think this patch can break compatibility with ARM (cc @Pennyzct  @Weichen81 )

**Comment by jcvenegas** [comment]
@wenlxie intesting what is the maximal amount of memory you can hotplug currently with out get the agent killed ?

**Comment by sboeuf** [comment]
@wenlxie the way memory hotplug works for ARM has been implemented [here](https://github.com/kata-containers/agent/pull/443) by @Pennyzct.

@Pennyzct could you take a look at this PR and provide some feedback since I'm worried having the config `CONFIG_MEMORY_HOTPLUG_DEFAULT_ONLINE` enabled might not be the right fix here.

**Comment by wenlxie** [comment]
@jcvenegas 
> @wenlxie intesting what is the maximal amount of memory you can hotplug currently with out get the agent killed ?

In my env,  the limitation is 120Gi.  I tested 119Gi memory, it is ok.

**Comment by wenlxie** [comment]
@devimc 
I think this is the reason.  https://lwn.net/Articles/668944/
 ```
Currently, all newly added memory blocks remain in 'offline' state unless
someone onlines them, some linux distributions carry special udev rules
like:

SUBSYSTEM=="memory", ACTION=="add", ATTR{state}=="offline", ATTR{state}="online"

to make this happen automatically. This is not a great solution for virtual
machines where memory hotplug is being used to address high memory pressure
situations as such onlining is slow and a userspace process doing this
(udev) has a chance of being killed by the OOM killer as it will probably
require to allocate some memory.

```

**Comment by Pennyzct** [comment]
Hi~@wenlxie @devimc @sboeuf 
For ARM, we also rely on `listenToUdevEvents` and `onlineCPUMem` to online hot-plugged memory blocks. 
And I have tried this `CONFIG_MEMORY_HOTPLUG_DEFAULT_ONLINE`, it works well, which means all above manual dependency also could be eliminated on arm platform. ;)

**Comment by jcvenegas** [comment]
@chavafg  I am afraid we dont have a full kata testing in this repository we need to add one to fully test.
@wenlxie meanwhile.

- Seems that CI SoB-check is failing due to missing commit body, please add something similar that you put in this PR. Also please open an github issue for this. And also add to the commit body a line like:
```
Fixes: #ISSUE_NUMBER
```
- While we have a job here to fully test kata we need to open a dummy in runtime and add a depends on I can do that for you.

**Comment by jcvenegas** [comment]
@wenlxie @sboeuf if this works well we will need to sync to also remove `onlineCPUMem` the memory part.

**Comment by sboeuf** [comment]
@jcvenegas 
Ok so if I summarize here, by having this flag, the kernel will take care of hotplugging any new memory added on the fly while the system is already booted, right?

@Pennyzct confirmed this feature works well for ARM and we don't need the manual logic of doing `OnlineCpuMem` from the agent, is that right?

We need to try this out with `x86_64` and `s390`, to validate that the manual online cpu and memory logic can be totally removed.
@jcvenegas can you check this for x86?
@alicefr can you check this for s390?

**Comment by amshinde** [comment]
@jcvenegas @chavafg Can we add full CI testing to this PR?
Would be good to test changes to be made in the agent (to remove manual online memory)  with this change.

**Comment by Pennyzct** [comment]
@sboeuf yep. that's what I mean. I deleted two parts to test this config, [one](https://github.com/kata-containers/agent/blob/master/agent.go#L582) in `listenToUdevEvents`, and the whole `onlineCPUMem`, and it works well. all memory blocks automatically online. ;)

**Comment by wenlxie** [comment]
> @chavafg I am afraid we dont have a full kata testing in this repository we need to add one to fully test.
> @wenlxie meanwhile.
> 
> * Seems that CI SoB-check is failing due to missing commit body, please add something similar that you put in this PR. Also please open an github issue for this. And also add to the commit body a line like:
> 
> ```
> Fixes: #ISSUE_NUMBER
> ```
> * While we have a job here to fully test kata we need to open a dummy in runtime and add a depends on I can do that for you.

@jcvenegas  Thanks, now the SoB check had passed.

**Comment by alicefr** [comment]
@wenlxie you can remove the option for s390x. The memory hotplug is not supported by qemu KVM yet

**Comment by wenlxie** [comment]
> @wenlxie you can remove the option for s390x. The memory hotplug is not supported by qemu KVM yet

@alicefr  removed, thanks

**Comment by wenlxie** [comment]
kata_config_version is also need to be changed. Or the Ci job will fail.

**Comment by wenlxie** [comment]
It seems CI has issues before, can someone help to re trigger this? 
Or maybe I can have the permission to do this :)

**Comment by jcvenegas** [comment]
@wenlxie `ERROR: Please bump version in kernel/kata_config_version` seems that some kernel changes were merged please rebase and bump again the kernel config version file.

**Comment by jodh-intel** [comment]
@wenlxie - the CI is failing as you need to update the kernel version file:

```
Kernel directory has changes check kernel/kata_config_version changed
ERROR: Please bump version in kernel/kata_config_version
```

**Comment by wenlxie** [comment]
@jodh-intel  thanks.
I had bump the kata_config_version

**Comment by wenlxie** [comment]
CI job need to be run again. Can anyone help to trigger it?

**Comment by jodh-intel** [comment]
Sure....

/retest

File: kernel/kata_config_version [hunk 1]
```diff
@@ -1 +1 @@
-25
+26
```
Comment (grahamwhaley) [inline kernel/kata_config_version]: Hi @wenlxie - sorry about this, but I think the kernel config version has moved ahead in the master branch again - can you rebase and repush... sorry for the hassle - let's see if we can get this one moving again...

**Comment by wenlxie** [comment]
@grahamwhaley I rebase it and need someone help to trigger the test, thank.

**Comment by wenlxie** [comment]
The kernel config version bumps again. :(

**Comment by jcvenegas** [comment]
@jodh-intel  could you merge to  another bump race condition :P  ?

**Comment by jodh-intel** [comment]
Restarted CI which failed due to a timeout.

**Comment by jcvenegas** [comment]
time out again copy here to check if is a not stable test
```
• [SLOW TEST:5.021 seconds]
docker exit code
/tmp/jenkins/workspace/kata-containers-packaging-ubuntu-18-04-PR/go/src/github.com/kata-containers/tests/integration/docker/exitcode_test.go:20
  check exit codes
  /tmp/jenkins/workspace/kata-containers-packaging-ubuntu-18-04-PR/go/src/github.com/kata-containers/tests/vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
    with exit code '256' when interactive mode is: 'false', it should exit '0'
    /tmp/jenkins/workspace/kata-containers-packaging-ubuntu-18-04-PR/go/src/github.com/kata-containers/tests/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:46
------------------------------
Build timed out (after 50 minutes). Marking the build as aborted.
```

**Comment by jodh-intel** [comment]
18.04 timed out. Prodded it with a sharper stick this time...

**Comment by amshinde** [comment]
@wenlxie This needs another update to `kata_config_version`! Can you please do that?
Would like to see this merged soon.

**Comment by wenlxie** [comment]
@amshinde  I rebased it. But the ci job always fails with timeout, no more infos to debug the issue.

**Comment by grahamwhaley** [comment]
/test
let's see how the CI is now - and keep an eye open to determine if this is a CI fail specific to this PR, or a more generic instability...
thx for your patience @wenlxie

**Comment by wenlxie** [comment]
> /test
> let's see how the CI is now - and keep an eye open to determine if this is a CI fail specific to this PR, or a more generic instability...
> thx for your patience @wenlxie

@grahamwhaley  @jodh-intel  Thanks

**Pull request merged**